### PR TITLE
Edges Chantelle Media Ranker

### PR DESCRIPTION
**Pull request by BASIC-Belic**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | The only custom method I created was in application record "self.format_date(given_date)" , which was a class method I could use sitewide to format the dates in a way I wanted. I also used rails convention to create a validation of vote-- that each work voted on has a unique user. 
Describe how you approached testing that model method. What edge cases did you come up with? | I didn't come up with any tests for it. Since it was display logic, I just saw that the things displayed the way I wanted them to. One special model test I did do was the validations to test  for vote uniqueness. I wanted to test the cases where it should be valid, should be invalid (ie invalid for missing user/vote, invalid for duplicate user, valid for unique user/vote). I didn't really come up with any edge cases. 
What are `session` and `flash`? What is the difference between them? |  Session and flash are rails hashlike objects that persist because of browser cookies. Session persists for as long as the user does not end the session. Flash lasts one extra request-response cycle, and only the current request-response cycle for flash.now.
Describe a controller filter you wrote. | I used a controller filter to reduce repetition-- in the case of finding an ActiveRecord model instance by params. I also used it to populate the instance variables @work_votes and @user_votes for the user and work show page-- reducing the amount of logic in those views as well as keeping the methods private for security purposes.
What was one thing that you gained more clarity on through this assignment? | Controller filters, sessions. 
What is the Heroku URL of your deployed application? | https://chantelle-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Comprehension questions	| x
**General** |
Rails fundamentals (RESTful routing, use of named paths) | x
Views are well-organized (DRY, use of semantic HTML, use of partials) | x
Errors are reported to the user | x
Business logic lives in the models | Chose to not have any custom logic in the models, and instead put it in the controllers
Models are thoroughly tested, including relations, validations and any custom logic | no tests on relations, no custom logic methods to test, but validations tests are very thorough!
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | x
The ID of the current user is stored in the session | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count | x
Splash page contains a media spotlight | x
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | x
Individual user pages and the user list are present | x
**Optional - Styling** |
Bootstrap is used appropriately | x
Look and feel is similar to the original | x
**Overall** |

Great job Chantelle! Your project submission looks great -- the code is good, and the deployed version is wonderful!

As always, I think your code is good, and I can tell that you are experimenting and trying different things. My comments on this code are mostly minor suggestions and questions.

Overall, good work!

File: app/controllers/welcome_controller.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class WelcomeController < ApplicationController
+
+  before_action :find_top_ten_movies, only: :index
+  before_action :find_top_ten_books, only: :index
+  before_action :find_top_ten_albums, only: :index
+  before_action :find_spotlight_work, only: :index
```
Comment (tildeee) [inline app/controllers/welcome_controller.rb:6]: If you're only using it for one action, is it worth making it a before action? Why not just put it in the index action? I know we taught to use the pattern of a before_action, but you can also execute your private methods by just calling them in the `index` method too ;)

File: app/controllers/welcome_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  def index
+  end
+
+  private
+
+  def find_spotlight_work
+  @spotlight_work = Work.left_joins(:votes).group(:id).order('COUNT(votes.id) DESC').limit(1)
```
Comment (tildeee) [inline app/controllers/welcome_controller.rb:14]: Interesting... I think there could be a better way to find the work with the most votes. Also, please watch your indentation

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Work < ApplicationRecord
+  VALID_CATEGORIES =  %w( movie book album )
```
Comment (tildeee) [inline app/models/work.rb:2]: Lovely use of a constant variable! This is my favorite!

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+  has_many :votes, dependent: :destroy
+  validates :title, presence: true, uniqueness: { scope: :category, message: "this work was already added for this category" }
```
Comment (tildeee) [inline app/models/work.rb:4]: Nicely done

File: app/views/sessions/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+
+<%= form_with method: :post, url: login_path do |f| %>
+<h2>Log in</h2>
+<p><%= f.label :username %>
+<%= f.text_field :username%>
+</p>
+<p><%= f.submit "Log in", class: "btn btn-primary" %></p>
+<% end %>
+<p>New user? <%= link_to "Sign up now!", signup_path %></p>
```
Comment (tildeee) [inline app/views/sessions/new.html.erb:9]: Why is there a sign up page? Was there a requirement for this? ;)

File: app/views/works/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+         <th scope="col" >Title</th>
+         <th scope="col" >Created By</th>
+         <th scope="col" width>Published</th>
+         <th scope="col" width>Upvote</th>
+       </tr>
+     </thead>
+
+     <tbody>
+       <% @works.each do |work| %>
+       <% if work.category == 'album' %>
```
Comment (tildeee) [inline app/views/works/index.html.erb:19]: Is there a way we could reduce the repeated code here that checks on the work's category? This is essentially duplicated code except for the h4 and the category.

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+   root 'welcome#index'
+   resources :works
+   resources :users, except: [:new, :edit, :update, :delete]
```
Comment (tildeee) [inline config/routes.rb:6]: With a list like this, you might want to use the "only" syntax to get index and show

File: test/models/user_test.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+require 'pry'
+
+describe User do
+
+  before do
+    @user = User.create!(username: "Julia")
+    @saved_user = users(:first)
+    #User.first
+  end
+
```
Comment (tildeee) [inline test/models/user_test.rb:12]: tests on relations to votes that users have?

File: test/models/user_test.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  describe 'validations' do
+    it 'is valid when a username is present' do
+
+      empty_hash = {}
+
+      expect(@saved_user.valid?).must_equal true
+      expect(@saved_user).must_be_kind_of User
+      expect(@saved_user.errors.messages).must_equal empty_hash
```
Comment (tildeee) [inline test/models/user_test.rb:20]: Is this helpful to have a variable named `empty_hash` whose value is just `{}`? It would be better to use the minitest matcher `must_be_empty` because it's more evocative and it isn't checking that it's literally an empty hash

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
+require "test_helper"
+
+require 'pry'
+
+describe User do
+
+  before do
+    @user = User.create!(username: "Julia")
```
Comment (tildeee) [inline test/models/user_test.rb:8]: Why make a user in the tests and why not start with a valid user defined in the fixtures?

File: test/models/user_test.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+      expect(@user.valid?).must_equal false
+      expect(@user.errors.messages).must_include :username
+    end
+
+    it 'is invalid if user of same username is created' do
+      #TO DO: populate her
+      @user.username = @saved_user.username
+
+      expect(@saved_user.username).must_equal 'first-name123'
+      expect(@user.username).must_equal 'first-name123'
```
Comment (tildeee) [inline test/models/user_test.rb:36]: This is nit-picking but: It's pretty clear that the usernames of these two users will be the same because of the line above (line 33 where you set them equal to each other). I don't think it's necessary to have these two `must_equal`s above

File: test/models/vote_test.rb [hunk 12]
```diff
@@ -0,0 +1,59 @@
+require "test_helper"
+
+describe Vote do
+
+  before do
+    @arbitrary_vote = Vote.new(user: User.find_by(username: "first-name123"), work: Work.find_by(category: 'movie'))
```
Comment (tildeee) [inline test/models/vote_test.rb:6]: Again, why not start with a valid vote defined in the fixtures? You follow the pattern of have a reference to a vote that is valid and then in the tests themselves, to test that they're invalid, you set them attribute by attribute to something invalid. So why not start with a valid fixture?

File: test/models/vote_test.rb [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    it 'is is invalid if user and work uninstantiated' do
+
+      @arbitrary_vote.user = nil
+      @arbitrary_vote.work = nil
+
+      expect(@arbitrary_vote.valid?).must_equal false
+      expect(@arbitrary_vote.errors.messages).must_include :user, :work
+    end
+
+    it 'is is invalid if user id and work id are nil' do
```
Comment (tildeee) [inline test/models/vote_test.rb:29]: how is this test different than the one above? If you had tests on relations, I would even argue this is even more redundant

File: test/models/work_test.rb [hunk 14]
```diff
@@ -0,0 +1,81 @@
+require "test_helper"
+
+describe Work do
+
+  before do
+    @work = Work.new(title: "a work", category: "book")
```
Comment (tildeee) [inline test/models/work_test.rb:6]: Again, why not start with a valid work defined in the fixtures?
